proof/refine/ARM/StateRelation.thy

```diff
@@ -880,33 +891,66 @@ lemma sc_replies_relation_prev_list:
   apply (clarsimp simp: sc_of_def opt_map_def projectKO_opt_sc split: option.splits)
   done
 
-(* FIXME: can generalise in terms of "sym heaps", maybe later *)
-lemma sym_refs_prev_eq_next_sc_replies:
-  "\<lbrakk> sym_refs (list_refs_of_replies' s');
-     heap_list (replyPrevs_of s') (scReply sc') replies\<rbrakk>
-    \<Longrightarrow> heap_list (replyNexts_of s') (scReply sc') (rev replies)"
-  oops
+lemma list_refs_of_replies'_reftype[simp]:
+  "(p, reftype) \<in> list_refs_of_replies' s' p' \<Longrightarrow> reftype \<in> {ReplyPrev, ReplyNext}"
+  by (clarsimp simp: list_refs_of_replies'_def list_refs_of_reply'_def get_refs_def2
+              split: option.split_asm)
+
+(* FIXME RT: move *)
+lemma opt_map_SomeD:
+  "f x = Some y \<Longrightarrow> (f |> g) x = g y"
+  by (clarsimp simp: opt_map_Some_def)
+
+lemma replyNext_replyNexts_of_opt_map[simp]:
+  "\<lbrakk>ksPSpace s' p = Some (KOReply reply); replyNext reply = Some (Next p')\<rbrakk>
+    \<Longrightarrow> (replyNexts_of s' |> replyPrevs_of s') p = replyPrevs_of s' p'"
+  by (clarsimp simp: opt_map_SomeD projectKO_opt_reply split: option.split)
+
+lemma sym_replies_next_then_prev_id:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    ksPSpace s' p = Some (KOReply reply); replyNext_of reply = Some p'\<rbrakk>
+   \<Longrightarrow> (replyNexts_of s' |> replyPrevs_of s') p = Some p"
+  supply opt_map_SomeD[simp]
+  apply (drule sym_refsD[rotated, where x=p and y=p' and tp=ReplyNext])
+   apply (clarsimp simp: map_set_def list_refs_of_reply'_def projectKO_opt_reply)
+  apply (clarsimp simp: map_set_def list_refs_of_reply'_def get_refs_def2
+                 split: option.splits)
+  done
 
-lemma sc_replies_relation_next_list:
-  "\<lbrakk> sc_replies_relation s s'; sym_refs (list_refs_of_replies' s');
-     kheap s x = Some (kernel_object.SchedContext sc n);
-     ksPSpace s' x = Some (KOSchedContext sc')\<rbrakk>
-    \<Longrightarrow> heap_list (replyNexts_of s') (scReply sc') (rev (sc_replies sc))"
-  apply (frule (2) sc_replies_relation_prev_list)
-  apply (case_tac "sc_replies sc"; simp)
-  apply (clarsimp simp: sc_replies_relation_def sc_replies_of_scs_def scs_of_kh_def map_project_def)
-  apply (drule_tac x=x and y="sc_replies sc" in spec2)
-  apply (clarsimp simp: sc_of_def opt_map_def projectKO_opt_sc)
-  oops
+lemma sym_replies_prev_then_next_id:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    ksPSpace s' p = Some (KOReply reply); replyPrev reply = Some p'\<rbrakk>
```

Comment:
Same as above, I'd write `replyPrevs_of s' p = Some p'`

Reply:
See the projection version below.